containers/source-build/source_build.py

```diff
@@ -0,0 +1,455 @@
+) -> None:
+    """Extract a blob member and rename it."""
+    # strip 3 components: ./blobs/sha256
+    tar_cmd = [
+        "tar",
+        "--extract",
+        "-C",
+        dest_dir,
+        "--strip-components",
+        "3",
+        "-f",
+        tar_archive,
+        member,
+    ]
+    log.debug("extract blob member %r", tar_cmd)
+    run(tar_cmd, check=True, cwd=work_dir)
+    shutil.move(f"{dest_dir}/{os.path.basename(member)}", f"{dest_dir}/{rename_to}")
+
+
+def prepare_base_image_sources(
+    image: str, work_dir: str, sib_dirs: SourceImageBuildDirectories
+) -> bool:
+    log = logging.getLogger("source-build.base-image-sources")
+    parts = urllib.parse.urlparse(f"docker://{image}")
+    if parts.netloc != "registry.access.redhat.com":
+        log.info(
+            "Image %s does not come from supported registry. Skip handling the sources from it.",
+            image,
+        )
+        return False
+
+    base_image_sources_dir = create_dir(work_dir, "base_image_sources")
+    base_sources_extraction_dir = create_dir(base_image_sources_dir, "extraction_dir")
+
+    source_image_name = image + "-source"
+
+    if not registry_has_image(source_image_name):
+        logger.warning(
+            "The registry does not have corresponding source image %s", source_image_name
+        )
+        return False
+
+    cmd = ["skopeo", "copy", f"docker://{source_image_name}", f"dir:{base_sources_extraction_dir}"]
+    log.info(
+        "Copy source image %s into directory %s",
+        source_image_name,
+        str(base_sources_extraction_dir),
+    )
+    try:
+        run(cmd, check=True)
+    except CalledProcessError:
+        log.warning("Failed to copy source image to local %s", source_image_name)
+        return False
+
+    # bsi reads source RPMs from this directory
+    bsi_rpms_dir = create_dir(base_image_sources_dir, "bsi_rpms_dir")
+    sib_dirs.rpm_dir = str(bsi_rpms_dir)  # save this directory for executing bsi
+
+    # bsi reads extra sources from this directory
+    # each source is in its own directory, for instance, subdir/source_a.tar.gz
+    extra_src_dir = create_dir(base_sources_extraction_dir, "extra_src_dir")
+
+    # extract layers, primarily they are RPMs
+    with open(f"{base_sources_extraction_dir}/manifest.json", "r") as f:
+        manifest_data = json.load(f)
+
+    gathered = False
+
+    for layer in manifest_data["layers"]:
+        digest = layer["digest"].split(":")[-1]
+        log.debug("untar layer %s", digest)
+
+        blob_member = ""
+        symlink_member = ""
+        with tarfile.open(f"{base_sources_extraction_dir}/{digest}", "r:gz") as tar:
+            for member in tar:
+                if member.isfile():
+                    blob_member = member.name
+                elif member.issym():
+                    symlink_member = member.name
+
+        if symlink_member.startswith("./rpm_dir/"):
```

Comment:
Are the file paths in the manifest guaranteed to start with `./`? This feels like we're relying on an implementation detail

Reply:
Yes, relying on bsi's implementation detail :) I did this to avoid unpacking all the SRPMs and possible extra sources with `tar xvf`.

Reply:
I was going to suggest something like this to make it more robust:
```python
if Path(symlink_member).parts[0] == "rpm_dir":
    ...
```
But we also rely on the leading `./` when extracting (`# strip 3 components: ./blobs/sha256`)

We could change the extraction to be done by python's tarfile library and then we wouldn't have to rely on the `./`. But for now, relying on the implementation detail is fine